Source/WebKit/UIProcess/WebProcessProxy.cpp

```diff
@@ -1073,6 +1073,15 @@ void WebProcessProxy::createGPUProcessConnection(IPC::Connection::Handle&& conne
     protectedProcessPool()->createGPUProcessConnection(*this, WTFMove(connectionIdentifier), WTFMove(parameters));
 }
 
+void WebProcessProxy::resetGPUProcess()
+{
+    WEBPROCESSPROXY_RELEASE_LOG(Process, "resetGPUProcess: webPage=%p, pageProxyID=%" PRIu64 ", webPageID=%" PRIu64 " hasReset=%d", &webPage, webPage.identifier().toUInt64(), webPage.webPageID().toUInt64(), m_hasResetGPUProcess);
+    if (!m_hasResetGPUProcess) {
+        m_hasResetGPUProcess = true;
+        protectedProcessPool()->resetGPUProcess();
```

Comment:
If we think that the issue is happening because the GPUProcess is hung, we may be able to adopt another approach. Instead of asking us to kill the GPUProcess, the WebProcess could ask the UIProcess to check if the GPUProcess is responsive (we already have logic to check that). Then if the UIProcess detects the GPUProcess is unresponsive, it would kill it (We likely already have logic to kill on unresponsive GPUProcess detection).

Then it would feel less hacky to me and there would be no need to only allow it once per WebProcess. This means we could recover from bad behavior more than once, per lifetime of a WebProcess.

Note that a WebProcess may live for a long time due to the process cache and may get reused for plenty of loads.

Reply:
The issue is not that the GPUProcess is truly hung, but instead that it's idle and not doing any work, while the WebProcess is blocked waiting for it.

Reply:
> The issue is not that the GPUProcess is truly hung, but instead that it's idle and not doing any work
Are we sure that the GPUProcess is not suspended?

Reply:
I just looked at the spindump and it doesn't appear the GPUProcess is suspended. It does look idle though, and not hung, weird. Must be some kind of IPC issue?

Reply:
Yeah, almost certainly some sort of race/logic bug in IPC.